libbeat/publisher/queue/diskqueue/segments.go

```diff
@@ -222,20 +333,110 @@ func (segment *queueSegment) getWriterWithRetry(
 	return file, err
 }
 
-func readSegmentHeader(in *os.File) (*segmentHeader, error) {
+// readSegmentHeaderWithFrameCount reads the header from the beginning
+// of the file at the given path. If the header's frameCount is 0
+// (whether because it is from an old version or because the segment
+// file was not closed cleanly), it attempts to calculate it manually
+// by scanning the file, and returns a struct with the "correct"
+// frame count.
+func readSegmentHeaderWithFrameCount(path string) (*segmentHeader, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"couldn't open segment file '%s': %w", path, err)
+	}
+	defer file.Close()
+	// Wrap the handle to retry non-fatal errors and always return the full
+	// requested data length if possible, then read the raw header.
+	reader := autoRetryReader{file}
+	header, err := readSegmentHeader(reader)
+	if err != nil {
+		return nil, err
+	}
+	// If the header has a positive frame count then there is
+	// no more work to do, so return immediately.
+	if header.frameCount > 0 {
+		return header, nil
+	}
+	// If we made it here, we loaded a valid header but the frame count is
+	// zero, so we need to check it with a manual scan.
```

Comment:
under which circumstances can the count be 0?

Reply:
Two possible cases:
- The segment is from the previous version which had no frameCount field, or
- The segment file was not closed cleanly during the previous session and still has the placeholder value

Reply:
Let's add the "why" to the code comment. The fact that the frame count is 0 and we do a scan is obvious from the code. The "why" we are in this state is not and should be documented.